### DISCUSSION TITLE
Need help to test pinia store using router

### DISCUSSION THREAD

**Discussion by jbaubree** [vuejs/pinia#2239] [Help and Questions] [answered]
I am using pinia with vite-ssg.
My pinia module is configured like this

```
import { createPinia } from 'pinia'
import type { UserModule } from '~/types'

export const pinia = createPinia()

export const install: UserModule = ({ isClient, initialState, app, router }) => {
  app.use(pinia)

  pinia.use(({ store }) => {
    store.router = markRaw(router)
  })

  if (isClient)
    pinia.state.value = (initialState.pinia) || {}

  else initialState.pinia = pinia.state.value
}
```

I can use this.router.push function in my store and all works perfectly.
I need to test with vitest this store so i use this test:
```
describe('user store', () => {
  beforeEach(() => {
    const pinia = createPinia()
    pinia.use(({ store }) => {
      store.router = markRaw(router)
    })
    setActivePinia(createPinia())
  })
  it('should logout', () => {
    const userStore = useUserStore()
    userStore.logout()
    expect(userStore.profile).toEqual(null)
  })
})

```

Unfortunately i get an error running test "Cannot read properties of undefined (reading 'push')" from this.router.push function.
Can someone help me to configure pinia.router in test mode ?

**Comment by ghost23** [comment, marked as answer]
Hi,

I just had the exact same problem. What solved it for me was to add a fake app (as is actually described in the Pinia documentation in the Testing section), like so:

```
beforeEach(() => {
  const pinia = createPinia();
  pinia.use(({ store }) => {
    store.router = markRaw(router);
  });
  const app = createApp({}); // <-- !
  app.use(pinia);            // <-- !
  setActivePinia(pinia);
})
```

Because according to the documentation without an app, plugins (in this case the router) will not actually be added to pinia.

**Reply by posva** [reply]
FYI: you can use the `fakeApp` option: https://pinia.vuejs.org/api/@pinia/testing/interfaces/TestingOptions.html#fakeApp-

**Comment by Natalia912** [comment]
I've encountered the same problem. For some reason creating fake app doesn't seem to work
